Examples/Io/Root/src/RootTrackParameterWriter.cpp

```diff
@@ -63,30 +114,106 @@ ActsExamples::ProcessCode ActsExamples::RootTrackParameterWriter::endRun() {
   }
   return ProcessCode::SUCCESS;
 }
 
 ActsExamples::ProcessCode ActsExamples::RootTrackParameterWriter::writeT(
     const ActsExamples::AlgorithmContext& ctx,
-    const std::vector<BoundTrackParameters>& trackParams) {
-  if (m_outputFile == nullptr)
+    const TrackParametersContainer& trackParams) {
+  using HitParticlesMap = IndexMultimap<ActsFatras::Barcode>;
+  using HitSimHitsMap = IndexMultimap<Index>;
+
+  if (m_outputFile == nullptr) {
     return ProcessCode::SUCCESS;
+  }
+
+  // Read additional input collections
+  const auto& trackParametersSeedMap =
+      ctx.eventStore.get<TrackParametersSeedMap>(
+          m_cfg.inputTrackParametersSeedMap);
+  const auto& seeds = ctx.eventStore.get<SimSeedContainer>(m_cfg.inputSeeds);
+  const auto& particles =
+      ctx.eventStore.get<SimParticleContainer>(m_cfg.inputParticles);
+  const auto& simHits = ctx.eventStore.get<SimHitContainer>(m_cfg.inputSimHits);
+  const auto& hitParticlesMap =
+      ctx.eventStore.get<HitParticlesMap>(m_cfg.inputMeasurementParticlesMap);
+  const auto& hitSimHitsMap =
+      ctx.eventStore.get<HitSimHitsMap>(m_cfg.inputMeasurementSimHitsMap);
 
   // Exclusive access to the tree while writing
   std::lock_guard<std::mutex> lock(m_writeMutex);
 
   // Get the event number
   m_eventNr = ctx.eventNumber;
 
-  for (auto& params : trackParams) {
-    m_d0 = params.parameters()[0];
-    m_z0 = params.parameters()[1];
-    m_phi = params.parameters()[2];
-    m_theta = params.parameters()[3];
-    m_qp = params.parameters()[4];
+  // Loop over the estimated track parameters
+  for (size_t iparams = 0; iparams < trackParams.size(); ++iparams) {
+    // The reference surface of the parameters, i.e. also the reference surface
+    // of the first space point
+    const auto& surface = trackParams[iparams].referenceSurface();
+    // The estimated bound parameters vector
+    const auto params = trackParams[iparams].parameters();
+    m_loc0 = params[Acts::eBoundLoc0];
+    m_loc1 = params[Acts::eBoundLoc1];
+    m_phi = params[Acts::eBoundPhi];
+    m_theta = params[Acts::eBoundTheta];
+    m_qop = params[Acts::eBoundQOverP];
+    m_time = params[Acts::eBoundTime];
+    m_p = std::abs(1.0 / m_qop);
+    m_pt = m_p * std::sin(m_theta);
+    m_eta = std::atanh(std::cos(m_theta));
+
+    // Get the seed via the estimated parameters to seed map
+    const auto& groupedSeedIdx = trackParametersSeedMap.at(iparams);
+    const auto& seed = seeds[groupedSeedIdx.regionIdx][groupedSeedIdx.seedIdx];
+    // check the seed quality
+    ProtoTrack ptrack{seed.sp()[0]->measurementIndex(),
+                      seed.sp()[1]->measurementIndex(),
+                      seed.sp()[2]->measurementIndex()};
+    std::vector<ParticleHitCount> particleHitCounts;
+    identifyContributingParticles(hitParticlesMap, ptrack, particleHitCounts);
+    m_truthMatched = false;
+    if (particleHitCounts.size() == 1) {
```

Comment:
Can it be larger than one and still be truth matched?

Reply:
In principle, yes. But since the seed has only three space point, we require that all the three space points of the seed must be matched to the same truth particle to be tagged as `truth-matched`.